lib/src/rules/no_wildcard_variable_uses.dart

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+import '../util/ascii_utils.dart';
+
+const _desc = r"Don't use wildcard parameters or variables.";
+
+const _details = r'''
+**DON'T** use wildcard parameters or variables.
+
+In a future version of the Dart langauge, wildcard parameters and local
+variables (e.g., named with just underscores like `_`, `__`, `___`, etc.) will
+become non-binding. When that happens, existing code that uses wildcard
+parameters or variables will break. In anticipation and to make adoption
+easier, this lint disallows wildcard and variable parameter uses.
+
+
+**BAD:**
+```dart
+var _ = 1;
+print(_); // LINT
+```
+
+```dart
+void f(int __) {
+  print(__); // LINT multiple underscores too
+}
+```
+
+**GOOD:**
+```dart
+for (var _ in [1, 2, 3]) count++;
+```
+
+```dart
+var [a, _, b, _] = [1, 2, 3, 4];
+```
+''';
+
+class NoWildcardVariableUses extends LintRule {
+  static const LintCode code = LintCode(
+      'no_wildcard_variable_uses', 'The referenced identifier is a wildcard.',
+      correctionMessage: 'Use an identifier name that is not a wildcard.');
+
+  NoWildcardVariableUses()
+      : super(
+            name: 'no_wildcard_variable_uses',
+            description: _desc,
+            details: _details,
+            group: Group.errors);
+
+  @override
+  LintCode get lintCode => code;
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this);
+    registry.addSimpleIdentifier(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitSimpleIdentifier(SimpleIdentifier node) {
+    if (node.name.isJustUnderscores) {
```

Comment:
I *think* it's just this simple... Please tell me if I'm over-simplifying though!

Reply:
The lint says: don't *use*.
Why not: don't *declare*?
There might be multiple uses, but they original reason is that something declared it.

Also, this visit method will fire on getters `print(this._)`.
Do we want this?

Reply:
> Why not: don't declare?

Read https://github.com/dart-lang/sdk/issues/51221.

> Do we want this?

No, the lint should only fire for uses of local variables (including parameters). We need tests that verify that it doesn't fire for members, including constructors and top-level functions, getters, etc.

Reply:
Thanks @bwilkerson. I'd forgotten about class members.

Reply:
Ah, of course I should have read the issue discussion.

Reply:
OK. Added executable element filtering (and tests).  Thanks!